syft/execution/plan.py

```diff
@@ -396,41 +504,67 @@ def __call__(self, *args, **kwargs):
           and use the result(s) to instantiate to appropriate placeholder.
         - Return the instantiation of all the output placeholders.
         """
-        if self.forward is not None:  # if not self.is_built:
+
+        def instantiate_input(input_placeholders, args):
+            """
+                Function to instantiate all the placeholders in order, even if in nested structures.
+
+                Args:
+                    input_placeholders: the input placeholders generated at build time.
+                    args: the arguments received on __call__
+            """
+
+            for arg in args:
+                if isinstance(arg, FrameworkTensor):
+                    placeholder = input_placeholders.pop(0)
+                    placeholder.instantiate(arg)
+                elif isinstance(arg, (list, tuple, set)):
+                    instantiate_input(input_placeholders, arg)
+                elif isinstance(arg, dict):
+                    instantiate_input(input_placeholders, list(arg.values()))
+
+        # If the function was not built, we can't make typechecking.
+        if not self.is_built:
             if self.include_state:
                 args = (*args, self.state)
             return self.forward(*args)
 
-        else:
-            # Instantiate all the input placeholders in the correct order
-
-            input_placeholders = sorted(self.find_placeholders("#input"), key=tag_sort("input"))
-            for placeholder, arg in zip(input_placeholders, args):
-                placeholder.instantiate(arg)
-
-            for i, action in enumerate(self.actions):
-                cmd, _self, args, kwargs, return_placeholder = (
-                    action.name,
-                    action.target,  # target is equivalent to the "self" in a method
-                    action.args,
-                    action.kwargs,
-                    action.return_ids,
-                )
-                if _self is None:
-                    response = eval(cmd)(*args, **kwargs)  # nosec
-                else:
-                    response = getattr(_self, cmd)(*args, **kwargs)
+        # Input validation based on build input.
+        self.validate_input(args, kwargs)
 
-                self.instantiate(return_placeholder, response)
+        # Instantiate all the input placeholders in the correct order
+        input_placeholders = sorted(self.find_placeholders("#input"), key=tag_sort("input"))
 
-            # This ensures that we return the output placeholder in the correct order
-            output_placeholders = sorted(self.find_placeholders("#output"), key=tag_sort("output"))
-            response = [p.child for p in output_placeholders]
+        if self.forward is not None:
+            if self.include_state:
+                args = (*args, self.state)
+            return self.forward(*args)
 
```

Comment:
I have a slight preference here for the if/else structure to better understand the branching logic